### PR TITLE
Test filtering by tags [v1]

### PR DESCRIPTION
**Pull request by clebergnu**
This adds support for test tags (to choose categorized tests) to the safeloader, loader, and also to the list and run commands.

It should be fully functional, and based on the discussion on v0 (#1621), it includes INSTRUMENTED tags that do not have `:avocado: tags=` set in the filtering process.  To disable that behavior, the `--filter-by-tags-ignore-empty` option was added.

Still missing here is full user facing documentation.  I'll include it based on feedback on the new behavior.

---

Other changes from v0 (#1621):
 * Docstring style changes
 * Fixed conditional block that was misplaced in another `if/else` block (about the removal of the `tags` from the test parameters)
 * Introduced many tests to the loader filter option
 * Introduced the `ignore_empty` option to `loader.filter_test_tags()`
 * Removed `FILTERED` "test type" and consequently from the `avocado list -V` output
 * [RFC] Tagged most INSTRUMENTED example tests

### CODE REVIEW

File: avocado/plugins/list.py [hunk 1]
```diff
@@ -76,26 +76,13 @@ def _get_test_matrix(self, test_suite):
...
-                    stats[type_label.lower()] += 1
-                    type_label = decorator(type_label)
-                    id_label = params['name']
+            if isinstance(cls, str):
+                cls = test.Test
+                id_label = params['name']
+            type_label = type_label_mapping[cls]
+            decorator = decorator_mapping[cls]
+            stats[type_label.lower()] += 1
+            type_label = decorator(type_label)
```
Comment (ldoktor) [inline avocado/plugins/list.py]: I don't really remember what was the problem with inherited tests, but you are right, all instrumented tests should be discovered as `string`. Btw shouldn't we rather use `basestring` to include unicode names?

File: avocado/plugins/list.py [hunk 2]
```diff
@@ -132,9 +132,8 @@ def configure(self, parser):
         :param parser: Main test runner parser.
         """
         parser = super(List, self).configure(parser)
-        parser.add_argument('keywords', type=str, default=[], nargs='*',
-                            help="List of paths, aliases or other "
-                            "keywords used to locate tests. "
+        parser.add_argument('reference', type=str, default=[], nargs='*',
+                            help="List of test references (aliases or paths). "
```
Comment (ldoktor) [inline avocado/plugins/list.py]: Not really related to this PR, but I don't see a problem with it... (maybe the first 2 commits could be even cherry-picked...)

File: avocado/core/loader.py [hunk 3]
```diff
@@ -430,8 +422,7 @@ def get_decorator_mapping():
                 test.MissingTest: output.TERM_SUPPORT.fail_header_str,
                 BrokenSymlink: output.TERM_SUPPORT.fail_header_str,
                 AccessDeniedPath: output.TERM_SUPPORT.fail_header_str,
-                test.Test: output.TERM_SUPPORT.healthy_str,
-                FilteredOut: output.TERM_SUPPORT.warn_header_str}
+                test.Test: output.TERM_SUPPORT.healthy_str}
```
Comment (ldoktor) [inline avocado/core/loader.py]: OK, let's remove dead code, we can always add new filtered type the way it's the most appropriate when we do have the filtering (btw we can filter by method names `avocado list examples/tests/gendata.py:bsod -V` but it does not show FILTERED either)

File: avocado/plugins/list.py (hunk 1 above)
Comment (clebergnu) [inline avocado/plugins/list.py]: Yep, good catch!

Comment (clebergnu) [inline avocado/plugins/list.py]: BTW, I've incorated this suggestion and moved it to a separate PR (#1633),

**Comment by clebergnu** [comment]
Closing in favor a v2.

**Comment by apahim** [comment]
[Tests tags for filtering](https://trello.com/c/ulTkVqwv/804-tests-tags-for-filtering)
